File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Peer/PeerCoordinator.cs
using System.Collections.Concurrent;
using FccDesktopAgent.Core.Config;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace FccDesktopAgent.Core.Peer;

/// <summary>
/// Manages peer directory, heartbeat exchange, and leadership claim evaluation.
/// </summary>
public interface IPeerCoordinator
{
    /// <summary>Build a health response from the local agent's current state.</summary>
    PeerHealthResponse BuildHealthResponse();

    /// <summary>Process an incoming heartbeat from a peer.</summary>
    PeerHeartbeatResponse HandleIncomingHeartbeat(PeerHeartbeatRequest request);

    /// <summary>Evaluate an incoming leadership claim.</summary>
    PeerLeadershipClaimResponse HandleLeadershipClaim(PeerLeadershipClaimRequest request);

    /// <summary>Send heartbeats to all known peers.</summary>
    Task SendHeartbeatToAllPeersAsync(CancellationToken ct);

    /// <summary>Returns a snapshot of the current peer directory.</summary>
    IReadOnlyDictionary<string, PeerState> GetPeerStates();

    /// <summary>Whether the current PRIMARY peer is suspected down.</summary>
    bool IsPrimarySuspected { get; }

    /// <summary>Direct health probe to a specific peer agent.</summary>
    Task<PeerHealthResponse?> DirectHealthProbeAsync(string agentId, CancellationToken ct);

    /// <summary>Evaluate all peers for suspect/confirmed-down status based on missed heartbeats.</summary>
    Task EvaluateSuspectsAsync(CancellationToken ct);

    /// <summary>Bootstrap the peer directory from cloud-provided peer entries, filtering out self.</summary>
    void InitializeFromConfig(IEnumerable<PeerDirectoryEntry> peerEntries);

    /// <summary>The local agent's current leader epoch.</summary>
    long LocalEpoch { get; set; }
}

/// <summary>
/// Entry from cloud config representing a peer in the site HA cluster.
/// </summary>
public sealed class PeerDirectoryEntry
{
    public string AgentId { get; set; } = string.Empty;
    public string DeviceClass { get; set; } = string.Empty;
    public string PeerApiBaseUrl { get; set; } = string.Empty;
    public string RoleCapability { get; set; } = string.Empty;
}

public sealed class PeerCoordinator : IPeerCoordinator
{
    private readonly IPeerHttpClient _peerClient;
    private readonly IOptionsMonitor<AgentConfiguration> _config;
    private readonly ILogger<PeerCoordinator> _logger;
    private readonly ConcurrentDictionary<string, PeerState> _peers = new();
    private readonly DateTimeOffset _startedAtUtc = DateTimeOffset.UtcNow;

    private static readonly string AgentVersion =
        typeof(PeerCoordinator).Assembly.GetName().Version?.ToString(3) ?? "0.1.0";

    public long LocalEpoch { get; set; }

    public PeerCoordinator(
        IPeerHttpClient peerClient,
        IOptionsMonitor<AgentConfiguration> config,
        ILogger<PeerCoordinator> logger)
    {
        _peerClient = peerClient;
        _config = config;
        _logger = logger;
    }

    public PeerHealthResponse BuildHealthResponse()
    {
        var cfg = _config.CurrentValue;
        return new PeerHealthResponse
        {
            AgentId = cfg.DeviceId,
            SiteCode = cfg.SiteId,
            CurrentRole = cfg.CurrentRole,
            LeaderEpoch = LocalEpoch,
            FccReachable = true, // TODO: wire to connectivity monitor
            UptimeSeconds = (long)(DateTimeOffset.UtcNow - _startedAtUtc).TotalSeconds,
            AppVersion = AgentVersion,
            HighWaterMarkSeq = 0, // TODO: wire to replication sequence
            ReportedAtUtc = DateTimeOffset.UtcNow,
        };
    }

    public PeerHeartbeatResponse HandleIncomingHeartbeat(PeerHeartbeatRequest request)
    {
        var cfg = _config.CurrentValue;

        _peers.AddOrUpdate(
            request.AgentId,
            _ => new PeerState
            {
                AgentId = request.AgentId,
                DeviceClass = request.DeviceClass,
                CurrentRole = request.CurrentRole,
                LeaderEpoch = request.LeaderEpoch,
                PeerApiBaseUrl = string.Empty, // Will be populated from config
                LastHeartbeatReceivedUtc = DateTimeOffset.UtcNow,
                ConsecutiveMissedHeartbeats = 0,
                SuspectStatus = SuspectStatus.Healthy,
                ReplicationLagSeconds = request.ReplicationLagSeconds,
            },
            (_, existing) =>
            {
                existing.CurrentRole = request.CurrentRole;
                existing.LeaderEpoch = request.LeaderEpoch;
                existing.DeviceClass = request.DeviceClass;
                existing.LastHeartbeatReceivedUtc = DateTimeOffset.UtcNow;
                existing.ConsecutiveMissedHeartbeats = 0;
                existing.SuspectStatus = SuspectStatus.Healthy;
                existing.ReplicationLagSeconds = request.ReplicationLagSeconds;
                return existing;
            });

        _logger.LogDebug("Heartbeat received from {AgentId} role={Role} epoch={Epoch}",
            request.AgentId, request.CurrentRole, request.LeaderEpoch);

        return new PeerHeartbeatResponse
        {
            AgentId = cfg.DeviceId,
            CurrentRole = cfg.CurrentRole,
            LeaderEpoch = LocalEpoch,
            Accepted = true,
            ReceivedAtUtc = DateTimeOffset.UtcNow,
        };
    }

    public PeerLeadershipClaimResponse HandleLeadershipClaim(PeerLeadershipClaimRequest request)
    {
        // Accept if proposed epoch is higher than our local epoch
        if (request.ProposedEpoch > LocalEpoch)
        {
            _logger.LogInformation(
                "Accepting leadership claim from {CandidateId} epoch={ProposedEpoch} (local={LocalEpoch})",
                request.CandidateAgentId, request.ProposedEpoch, LocalEpoch);

            LocalEpoch = request.ProposedEpoch;

            return new PeerLeadershipClaimResponse
            {
                Accepted = true,
                CurrentEpoch = LocalEpoch,
            };
        }

        _logger.LogInformation(
            "Rejecting leadership claim from {CandidateId} epoch={ProposedEpoch} (local={LocalEpoch})",
            request.CandidateAgentId, request.ProposedEpoch, LocalEpoch);

        return new PeerLeadershipClaimResponse
        {
            Accepted = false,
            Reason = $"Local epoch {LocalEpoch} >= proposed epoch {request.ProposedEpoch}",
            CurrentEpoch = LocalEpoch,
        };
    }

    public async Task SendHeartbeatToAllPeersAsync(CancellationToken ct)
    {
        var cfg = _config.CurrentValue;
        var heartbeat = new PeerHeartbeatRequest
        {
            AgentId = cfg.DeviceId,
            SiteCode = cfg.SiteId,
            CurrentRole = cfg.CurrentRole,
            LeaderEpoch = LocalEpoch,
            DeviceClass = "desktop",
            AppVersion = AgentVersion,
            UptimeSeconds = (long)(DateTimeOffset.UtcNow - _startedAtUtc).TotalSeconds,
            SentAtUtc = DateTimeOffset.UtcNow,
        };

        foreach (var (agentId, peer) in _peers)
        {
            if (string.IsNullOrEmpty(peer.PeerApiBaseUrl))
                continue;

            try
            {
                var response = await _peerClient.SendHeartbeatAsync(peer.PeerApiBaseUrl, heartbeat, ct);
                if (response is not null)
                {
                    peer.LastHeartbeatReceivedUtc = DateTimeOffset.UtcNow;
                    peer.ConsecutiveMissedHeartbeats = 0;
                    peer.CurrentRole = response.CurrentRole;
                    peer.LeaderEpoch = response.LeaderEpoch;
                    peer.SuspectStatus = SuspectStatus.Healthy;
                }
                else
                {
                    peer.ConsecutiveMissedHeartbeats++;
                    _logger.LogWarning("Heartbeat to {AgentId} returned null (missed={Missed})",
                        agentId, peer.ConsecutiveMissedHeartbeats);
                }
            }
            catch (Exception ex)
            {
                peer.ConsecutiveMissedHeartbeats++;
                _logger.LogWarning(ex, "Failed to send heartbeat to {AgentId} at {Url} (missed={Missed})",
                    agentId, peer.PeerApiBaseUrl, peer.ConsecutiveMissedHeartbeats);
            }
        }
    }

    public IReadOnlyDictionary<string, PeerState> GetPeerStates() =>
        new Dictionary<string, PeerState>(_peers);

    public bool IsPrimarySuspected
    {
        get
        {
            foreach (var (_, peer) in _peers)
            {
                if (peer.CurrentRole == "PRIMARY" &&
                    peer.SuspectStatus is SuspectStatus.Suspected or SuspectStatus.ConfirmedDown)
                {
                    return true;
                }
            }

            return false;
        }
    }

    public async Task<PeerHealthResponse?> DirectHealthProbeAsync(string agentId, CancellationToken ct)
    {
        if (!_peers.TryGetValue(agentId, out var peer) || string.IsNullOrEmpty(peer.PeerApiBaseUrl))
        {
            _logger.LogWarning("Cannot probe unknown peer {AgentId}", agentId);
            return null;
        }

        try
        {
            return await _peerClient.GetHealthAsync(peer.PeerApiBaseUrl, ct);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Direct health probe to {AgentId} failed", agentId);
            return null;
        }
    }

    public async Task EvaluateSuspectsAsync(CancellationToken ct)
    {
        var cfg = _config.CurrentValue;
        var heartbeatInterval = TimeSpan.FromSeconds(cfg.HeartbeatIntervalSeconds);
        // Number of missed heartbeats before suspicion (failoverTimeout / heartbeatInterval)
        var missedThreshold = cfg.FailoverTimeoutSeconds / cfg.HeartbeatIntervalSeconds;

        foreach (var (agentId, peer) in _peers)
        {
            if (peer.CurrentRole != "PRIMARY")
                continue;

            var elapsed = DateTimeOffset.UtcNow - peer.LastHeartbeatReceivedUtc;
            var missedCount = (int)(elapsed / heartbeatInterval);
            peer.ConsecutiveMissedHeartbeats = Math.Max(peer.ConsecutiveMissedHeartbeats, missedCount);

            if (peer.ConsecutiveMissedHeartbeats < missedThreshold)
            {
                if (peer.SuspectStatus != SuspectStatus.Healthy)
                {
                    _logger.LogInformation("Peer {AgentId} suspicion cleared", agentId);
                    peer.SuspectStatus = SuspectStatus.Healthy;
                }
                continue;
            }

            if (peer.SuspectStatus == SuspectStatus.Healthy)
            {
                peer.SuspectStatus = SuspectStatus.Suspected;
                _logger.LogWarning("Peer {AgentId} suspected down (missed={Missed}, threshold={Threshold})",
                    agentId, peer.ConsecutiveMissedHeartbeats, missedThreshold);
            }

            if (peer.SuspectStatus == SuspectStatus.Suspected)
            {
                // Perform a direct health probe to confirm
                var health = await DirectHealthProbeAsync(agentId, ct);
                if (health is not null)
                {
                    // Peer responded — clear suspicion
                    peer.SuspectStatus = SuspectStatus.Healthy;
                    peer.ConsecutiveMissedHeartbeats = 0;
                    peer.LastHeartbeatReceivedUtc = DateTimeOffset.UtcNow;
                    _logger.LogInformation("Peer {AgentId} responded to health probe — suspicion cleared", agentId);
                }
                else
                {
                    // Probe failed — confirm down
                    peer.SuspectStatus = SuspectStatus.ConfirmedDown;
                    _logger.LogWarning("Peer {AgentId} confirmed down after failed health probe", agentId);
                }
            }
        }
    }

    public void InitializeFromConfig(IEnumerable<PeerDirectoryEntry> peerEntries)
    {
        var cfg = _config.CurrentValue;
        var selfId = cfg.DeviceId;

        foreach (var entry in peerEntries)
        {
            // Filter out self
            if (string.Equals(entry.AgentId, selfId, StringComparison.OrdinalIgnoreCase))
                continue;

            _peers.AddOrUpdate(
                entry.AgentId,
                _ => new PeerState
                {
                    AgentId = entry.AgentId,
                    DeviceClass = entry.DeviceClass,
                    PeerApiBaseUrl = entry.PeerApiBaseUrl,
                    CurrentRole = "UNKNOWN",
                    LastHeartbeatReceivedUtc = DateTimeOffset.MinValue,
                    SuspectStatus = SuspectStatus.Healthy,
                },
                (_, existing) =>
                {
                    existing.DeviceClass = entry.DeviceClass;
                    existing.PeerApiBaseUrl = entry.PeerApiBaseUrl;
                    return existing;
                });

            _logger.LogInformation("Peer registered: {AgentId} class={Class} url={Url}",
                entry.AgentId, entry.DeviceClass, entry.PeerApiBaseUrl);
        }
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/peer/PeerCoordinator.kt
package com.fccmiddleware.edge.peer

import com.fccmiddleware.edge.config.ConfigManager
import com.fccmiddleware.edge.config.PeerDirectoryEntryDto
import com.fccmiddleware.edge.logging.AppLogger
import com.fccmiddleware.edge.logging.StructuredFileLogger
import com.fccmiddleware.edge.security.EncryptedPrefsManager
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.concurrent.ConcurrentHashMap

/**
 * PeerCoordinator — manages HA peer state, heartbeat exchange, and leadership tracking.
 *
 * Responsibilities:
 * - Maintains an in-memory directory of known peers and their health status.
 * - Builds health responses for incoming GET /peer/health requests.
 * - Processes incoming heartbeats and leadership claims from peers.
 * - Sends outbound heartbeats to all known peers on each cadence tick.
 * - Evaluates peers for suspect/down classification based on missed heartbeats.
 * - Provides [isPrimarySuspected] for the election coordinator to trigger failover.
 *
 * Thread safety: peer directory is a [ConcurrentHashMap]; individual fields on
 * [PeerState] are replaced atomically via copy-on-write (data class copy).
 */
class PeerCoordinator(
    private val peerHttpClient: PeerHttpClient,
    private val configManager: ConfigManager,
    private val encryptedPrefsManager: EncryptedPrefsManager,
    private val fileLogger: StructuredFileLogger,
) {

    companion object {
        private const val TAG = "PeerCoordinator"
        /** Number of missed heartbeats before a peer is SUSPECTED. */
        private const val SUSPECT_THRESHOLD = 3
        /** Number of missed heartbeats before a peer is CONFIRMED_DOWN. */
        private const val DOWN_THRESHOLD = 6
    }

    /** In-memory directory of known peers keyed by agentId. */
    val peers: ConcurrentHashMap<String, PeerState> = ConcurrentHashMap()

    /** The local agent's current HA role. Updated from config and election results. */
    @Volatile
    var currentRole: String = "STANDBY_HOT"

    /** The current leader epoch. Updated from heartbeats and election results. */
    @Volatile
    var leaderEpoch: Long = 0

    /** The agent ID of the current leader. */
    @Volatile
    var leaderAgentId: String? = null

    /** Monotonic sequence for the local agent's high-water-mark. */
    @Volatile
    var highWaterMarkSeq: Long = 0

    /** Service start time for uptime reporting. */
    private val startedAtMs = System.currentTimeMillis()

    /** Whether this coordinator has been initialized from config. */
    @Volatile
    var initialized = false
        private set

    // -------------------------------------------------------------------------
    // Initialization
    // -------------------------------------------------------------------------

    /**
     * Initialize peer directory and local role from the current site HA config.
     * Called once during startup after config is loaded.
     */
    fun initializeFromConfig() {
        val config = configManager.config.value ?: return
        val haConfig = config.siteHa
        if (!haConfig.enabled) return

        currentRole = haConfig.currentRole
        leaderEpoch = haConfig.leaderEpoch
        leaderAgentId = haConfig.leaderAgentId

        val localAgentId = encryptedPrefsManager.deviceId ?: return

        for (entry in haConfig.peerDirectory) {
            if (entry.agentId == localAgentId) continue
            val baseUrl = entry.peerApiBaseUrl ?: buildBaseUrl(entry)
            peers[entry.agentId] = PeerState(
                agentId = entry.agentId,
                deviceClass = entry.deviceClass,
                currentRole = entry.currentRole,
                leaderEpoch = entry.leaderEpochSeen ?: haConfig.leaderEpoch,
                peerApiBaseUrl = baseUrl,
            )
        }

        initialized = true
        AppLogger.i(TAG, "Initialized with ${peers.size} peer(s), role=$currentRole, epoch=$leaderEpoch")
    }

    // -------------------------------------------------------------------------
    // Health response (GET /peer/health)
    // -------------------------------------------------------------------------

    /**
     * Build a health response for incoming peer health checks.
     */
    fun buildHealthResponse(): PeerHealthResponse {
        val config = configManager.config.value
        val agentId = encryptedPrefsManager.deviceId ?: "unknown"
        val siteCode = encryptedPrefsManager.siteCode ?: "unknown"
        val fccReachable = true // Simplified — real implementation checks FCC adapter
        val uptimeSeconds = (System.currentTimeMillis() - startedAtMs) / 1000L
        val appVersion = config?.rollout?.minAgentVersion ?: "1.0.0"

        return PeerHealthResponse(
            agentId = agentId,
            siteCode = siteCode,
            currentRole = currentRole,
            leaderEpoch = leaderEpoch,
            fccReachable = fccReachable,
            uptimeSeconds = uptimeSeconds,
            appVersion = appVersion,
            highWaterMarkSeq = highWaterMarkSeq,
            reportedAtUtc = Instant.now().toString(),
        )
    }

    // -------------------------------------------------------------------------
    // Incoming heartbeat (POST /peer/heartbeat)
    // -------------------------------------------------------------------------

    /**
     * Process a heartbeat received from a peer.
     * Updates the peer directory and returns a response.
     */
    fun handleIncomingHeartbeat(request: PeerHeartbeatRequest): PeerHeartbeatResponse {
        val localAgentId = encryptedPrefsManager.deviceId ?: "unknown"
        val localSiteCode = encryptedPrefsManager.siteCode ?: "unknown"

        // Validate site code
        if (request.siteCode != localSiteCode) {
            AppLogger.w(TAG, "Heartbeat rejected: site code mismatch (${request.siteCode} != $localSiteCode)")
            return PeerHeartbeatResponse(
                agentId = localAgentId,
                currentRole = currentRole,
                leaderEpoch = leaderEpoch,
                accepted = false,
                receivedAtUtc = Instant.now().toString(),
            )
        }

        // Update peer state
        val now = Instant.now()
        peers.compute(request.agentId) { _, existing ->
            (existing ?: PeerState(
                agentId = request.agentId,
                deviceClass = request.deviceClass,
                currentRole = request.currentRole,
                leaderEpoch = request.leaderEpoch,
                peerApiBaseUrl = null,
            )).copy(
                currentRole = request.currentRole,
                leaderEpoch = request.leaderEpoch,
                lastHeartbeatReceivedAt = now,
                consecutiveMissedHeartbeats = 0,
                suspectStatus = SuspectStatus.HEALTHY,
                replicationLagSeconds = request.replicationLagSeconds,
            )
        }

        // Accept epoch advancement from heartbeat
        if (request.leaderEpoch > leaderEpoch) {
            leaderEpoch = request.leaderEpoch
            leaderAgentId = request.leaderAgentId
            AppLogger.i(TAG, "Epoch advanced to ${request.leaderEpoch} via heartbeat from ${request.agentId}")
        }

        return PeerHeartbeatResponse(
            agentId = localAgentId,
            currentRole = currentRole,
            leaderEpoch = leaderEpoch,
            accepted = true,
            receivedAtUtc = now.toString(),
        )
    }

    // -------------------------------------------------------------------------
    // Incoming leadership claim (POST /peer/claim-leadership)
    // -------------------------------------------------------------------------

    /**
     * Process a leadership claim from a candidate peer.
     *
     * Acceptance rules:
     * - Proposed epoch must be strictly greater than current epoch.
     * - If local agent is PRIMARY with same epoch, reject (split-brain prevention).
     */
    fun handleLeadershipClaim(request: PeerLeadershipClaimRequest): PeerLeadershipClaimResponse {
        val localSiteCode = encryptedPrefsManager.siteCode ?: "unknown"

        // Site code must match
        if (request.siteCode != localSiteCode) {
            return PeerLeadershipClaimResponse(
                accepted = false,
                reason = "Site code mismatch",
                currentEpoch = leaderEpoch,
            )
        }

        // Epoch must be strictly greater
        if (request.proposedEpoch <= leaderEpoch) {
            return PeerLeadershipClaimResponse(
                accepted = false,
                reason = "Proposed epoch ${request.proposedEpoch} <= current $leaderEpoch",
                currentEpoch = leaderEpoch,
            )
        }

        // If we are PRIMARY at the current epoch, reject to prevent split-brain
        if (currentRole == "PRIMARY" && request.proposedEpoch == leaderEpoch + 1) {
            // We're still alive as primary — reject unless we've already stepped down
            val localAgentId = encryptedPrefsManager.deviceId
            if (localAgentId != null && localAgentId != request.candidateAgentId) {
                AppLogger.w(TAG, "Rejecting claim from ${request.candidateAgentId}: we are still PRIMARY at epoch $leaderEpoch")
                return PeerLeadershipClaimResponse(
                    accepted = false,
                    reason = "Current PRIMARY is still alive",
                    currentEpoch = leaderEpoch,
                )
            }
        }

        // Accept the claim
        leaderEpoch = request.proposedEpoch
        leaderAgentId = request.candidateAgentId
        AppLogger.i(TAG, "Accepted leadership claim from ${request.candidateAgentId} at epoch ${request.proposedEpoch}")

        return PeerLeadershipClaimResponse(
            accepted = true,
            reason = null,
            currentEpoch = leaderEpoch,
        )
    }

    // -------------------------------------------------------------------------
    // Outbound heartbeat (called by CadenceController each tick)
    // -------------------------------------------------------------------------

    /**
     * Send heartbeats to all known peers. Updates suspect status for peers
     * that fail to respond. Called from the cadence loop.
     */
    fun sendHeartbeatToAllPeers() {
        if (!initialized) return
        val localAgentId = encryptedPrefsManager.deviceId ?: return
        val localSiteCode = encryptedPrefsManager.siteCode ?: return
        val config = configManager.config.value ?: return
        val uptimeSeconds = (System.currentTimeMillis() - startedAtMs) / 1000L

        val request = PeerHeartbeatRequest(
            agentId = localAgentId,
            siteCode = localSiteCode,
            currentRole = currentRole,
            leaderEpoch = leaderEpoch,
            leaderAgentId = leaderAgentId,
            configVersion = config.configId,
            replicationLagSeconds = 0.0,
            lastSequenceApplied = highWaterMarkSeq,
            deviceClass = config.identity.deviceClass,
            appVersion = config.rollout.minAgentVersion,
            uptimeSeconds = uptimeSeconds,
            sentAtUtc = OffsetDateTime.now(ZoneOffset.UTC).toString(),
        )

        for ((agentId, peerState) in peers) {
            val baseUrl = peerState.peerApiBaseUrl ?: continue
            val response = peerHttpClient.sendHeartbeat(baseUrl, request)
            if (response != null && response.accepted) {
                // Heartbeat acknowledged — peer is alive
                peers.computeIfPresent(agentId) { _, existing ->
                    existing.copy(
                        currentRole = response.currentRole,
                        leaderEpoch = response.leaderEpoch,
                    )
                }
            } else {
                // Heartbeat failed — increment missed count
                peers.computeIfPresent(agentId) { _, existing ->
                    val newMissed = existing.consecutiveMissedHeartbeats + 1
                    existing.copy(
                        consecutiveMissedHeartbeats = newMissed,
                        suspectStatus = when {
                            newMissed >= DOWN_THRESHOLD -> SuspectStatus.CONFIRMED_DOWN
                            newMissed >= SUSPECT_THRESHOLD -> SuspectStatus.SUSPECTED
                            else -> SuspectStatus.HEALTHY
                        },
                    )
                }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Suspect evaluation
    // -------------------------------------------------------------------------

    /**
     * Whether the current PRIMARY peer is suspected or confirmed down.
     * Used by [ElectionCoordinator] to decide whether to trigger failover.
     */
    val isPrimarySuspected: Boolean
        get() {
            val primaryAgentId = leaderAgentId ?: return false
            val localAgentId = encryptedPrefsManager.deviceId
            if (primaryAgentId == localAgentId) return false // We are primary
            val primaryState = peers[primaryAgentId] ?: return true // Unknown primary = suspected
            return primaryState.suspectStatus != SuspectStatus.HEALTHY
        }

    /**
     * Direct health probe to a specific peer (bypasses heartbeat cadence).
     * Used for confirming a suspect before triggering election.
     *
     * @return true if the peer responds to a health check, false otherwise.
     */
    fun directHealthProbe(agentId: String): Boolean {
        val peerState = peers[agentId] ?: return false
        val baseUrl = peerState.peerApiBaseUrl ?: return false
        val health = peerHttpClient.getHealth(baseUrl)
        return health != null
    }

    /**
     * Evaluate all peers and update their suspect status based on missed heartbeats.
     * Called periodically alongside heartbeat sending.
     */
    fun evaluateSuspects() {
        for ((agentId, peerState) in peers) {
            val lastHb = peerState.lastHeartbeatReceivedAt ?: continue
            val config = configManager.config.value ?: continue
            val heartbeatInterval = config.siteHa.heartbeatIntervalSeconds.toLong()
            val elapsedSeconds = java.time.Duration.between(lastHb, Instant.now()).seconds

            if (elapsedSeconds > heartbeatInterval * DOWN_THRESHOLD) {
                peers.computeIfPresent(agentId) { _, existing ->
                    if (existing.suspectStatus != SuspectStatus.CONFIRMED_DOWN) {
                        AppLogger.w(TAG, "Peer $agentId CONFIRMED_DOWN (${elapsedSeconds}s since last heartbeat)")
                        fileLogger.w(TAG, "Peer $agentId CONFIRMED_DOWN (${elapsedSeconds}s since last heartbeat)")
                    }
                    existing.copy(suspectStatus = SuspectStatus.CONFIRMED_DOWN)
                }
            } else if (elapsedSeconds > heartbeatInterval * SUSPECT_THRESHOLD) {
                peers.computeIfPresent(agentId) { _, existing ->
                    if (existing.suspectStatus == SuspectStatus.HEALTHY) {
                        AppLogger.w(TAG, "Peer $agentId SUSPECTED (${elapsedSeconds}s since last heartbeat)")
                    }
                    existing.copy(suspectStatus = SuspectStatus.SUSPECTED)
                }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private fun buildBaseUrl(entry: PeerDirectoryEntryDto): String? {
        val host = entry.peerApiAdvertisedHost ?: return null
        val port = entry.peerApiPort ?: return null
        val scheme = if (entry.peerApiTlsEnabled) "https" else "http"
        return "$scheme://$host:$port"
    }
}
